docs/winnt/cross-build-for-mingw64.md

```diff
@@ -0,0 +1,173 @@
+    mingw-w64           # cross-compiler for Win32           \
+    pkg-config           # needed to explicitly pull on Ubuntu
+```
+
+To keep everything isolated, and not muddy the installed system, we'll setup base location for the cross-built software `MINGW_ROOT` and set our host build platform as  `HOST_PLATFORM`
+
+```
+export MINGW_ROOT=/mingw64
+sudo mkdir $MINGW_ROOT
+sudo chown $USER:$USER $MINGW_ROOT
+
+# pick 32-bit or 64-bit
+export HOST_PLATFORM='i686_w64-mingw32'
+export HOST_PLATFORM='x86_64-w64-mingw32'
+```
+
+The sudo and chown are only needed if you are sticking it somewhere like the root directory. The first `HOST_PLATFORM` is for a 32-bit build, the second for 64-bit. Define only one.
+
+These triplets can be read as architecture (i686 or x86_64) -- system (w64 for minGW-w64) -- API (mingw's version of the `Win32` API). I'll assume these are setup (`source env.sh` with the scripts provided).
+
+## Cross-compiling NCurses Library
+
+The only outside dependency we need is [ncurses](https://invisible-island.net/ncurses/). Use the latest stable release. Support specifically targeted for MinGW-based ports was added in 5.8. These provide all the features of a normal terminal inside of standard Windows Console .The latest stable release is best here (6.3 as of writing).
+
+There is a port of ncurses called PDCurses that was designed to specifically support DOS and Windows, but it missing a sizable portion of widely used features. The general consensus is not use it if possible.
+
+Grab the source for ncurses, [*https://ftp.gnu.org/pub/gnu/ncurses/ncurses-6.3.tar.gz*](https://ftp.gnu.org/pub/gnu/ncurses/ncurses-6.3.tar.gz). Extract into a directory of your choosing, change into the that directory. Configure with these options:
+
+```
+./configure \
+	--host=$HOST_PLATFORM \
+	--prefix=$MINGW_ROOT \
+	--disable-getcap \
+	--disable-hard-tabs \
+	--disable-home-terminfo \
+	--disable-leaks \
+	--disable-macros \
+	--disable-overwrite \
+	--disable-shared
+	--disable-termcap \
+	--enable-assertions \
+	--enable-database \
+	--enable-echo   \
+	--enable-exp-win32 \
+	--enable-ext-funcs \
+	--enable-interop \
+	--enable-opaque-curses \
+	--enable-opaque-form \
+	--enable-opaque-menu \
+	--enable-opaque-panel \
+	--enable-pc-files \
+	--enable-sp-funcs \
+	--enable-static\
+	--enable-term-driver \
+	--enable-warnings \
+	--enable-widec \
+	--with-fallbacks=ms-terminal \
+	--with-normal \
+	--with-progs \
+	--without-ada \
+	--without-debug \
+	--without-libtool \
+	--without-manpages \
+	--without-shared \
+```
+This configuration is based on the recommended ones for compilation to Windows from NCurses developers ( [README.MinGW](https://github.com/mirror/ncurses/blob/master/README.MinGW) with the addition of `--enable-widec` to provide Unicode support. Script `build_ncurses.sh` is provided to configure with these options.
+
+Standard `make` and `make install` to build the library and install it and its headers into `$MINGW_ROOT`
```

Comment:
It seems `make install` in this step tries to unlink an existing global system installation of `ncurses`:

```
$ make install
( cd include && make DESTDIR="" RPATH_LIST="/mingw-root/build-root/winnt/lib" install )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/include“ wird betreten
mkdir -p /mingw-root/build-root/winnt/include/ncursesw
installing curses.h in /mingw-root/build-root/winnt/include/ncursesw
installing unctrl.h in /mingw-root/build-root/winnt/include/ncursesw
installing ncurses_dll.h in /mingw-root/build-root/winnt/include/ncursesw
installing term.h in /mingw-root/build-root/winnt/include/ncursesw
installing termcap.h in /mingw-root/build-root/winnt/include/ncursesw
installing ./tic.h in /mingw-root/build-root/winnt/include/ncursesw
installing ./term_entry.h in /mingw-root/build-root/winnt/include/ncursesw
installing ./nc_tparm.h in /mingw-root/build-root/winnt/include/ncursesw
installing ./win32_curses.h in /mingw-root/build-root/winnt/include/ncursesw
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/include“ wird verlassen
( cd ncurses && make DESTDIR="" RPATH_LIST="/mingw-root/build-root/winnt/lib" install )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/ncurses“ wird betreten
mkdir -p /mingw-root/build-root/winnt/lib
installing ../lib/libncursesw.a as /mingw-root/build-root/winnt/lib/libncursesw.a
/usr/bin/install -c -m 644 ../lib/libncursesw.a /mingw-root/build-root/winnt/lib/libncursesw.a
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/ncurses“ wird verlassen
( cd progs && make DESTDIR="" RPATH_LIST="/mingw-root/build-root/winnt/lib" install )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/progs“ wird betreten
mkdir -p /mingw-root/build-root/winnt/bin
/usr/bin/install -c -s tic.exe     /mingw-root/build-root/winnt/bin/`echo tic.exe|       sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'`
/usr/bin/install -c -s toe.exe     /mingw-root/build-root/winnt/bin/`echo toe.exe|       sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'`
linking infotocap.exe to tic.exe
( cd /mingw-root/build-root/winnt/bin && ln -s -f `echo tic.exe|       sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'` `echo infotocap.exe| sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'` )
linking captoinfo.exe to tic.exe
( cd /mingw-root/build-root/winnt/bin && ln -s -f `echo tic.exe|       sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'` `echo captoinfo.exe| sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'` )
/usr/bin/install -c -s infocmp.exe /mingw-root/build-root/winnt/bin/`echo infocmp.exe|   sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'`
/usr/bin/install -c -s clear.exe   /mingw-root/build-root/winnt/bin/`echo clear.exe|     sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'`
/usr/bin/install -c -s tabs.exe    /mingw-root/build-root/winnt/bin/`echo tabs.exe|      sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'`
/usr/bin/install -c -s tput.exe    /mingw-root/build-root/winnt/bin/`echo tput.exe|      sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'`
/usr/bin/install -c -s tset.exe    /mingw-root/build-root/winnt/bin/`echo tset.exe|      sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'`
linking reset.exe to tset.exe
( cd /mingw-root/build-root/winnt/bin && ln -s -f `echo tset.exe|      sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'` `echo reset.exe|     sed 's/.exe$//'|sed 's,x,x,'|sed 's/$/.exe/'` )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/progs“ wird verlassen
( cd panel && make DESTDIR="" RPATH_LIST="/mingw-root/build-root/winnt/lib" install )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/panel“ wird betreten
installing ../lib/libpanelw.a as /mingw-root/build-root/winnt/lib/libpanelw.a
/usr/bin/install -c -m 644 ../lib/libpanelw.a /mingw-root/build-root/winnt/lib/libpanelw.a
installing ./panel.h in /mingw-root/build-root/winnt/include/ncursesw
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/panel“ wird verlassen
( cd menu && make DESTDIR="" RPATH_LIST="/mingw-root/build-root/winnt/lib" install )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/menu“ wird betreten
installing ../lib/libmenuw.a as /mingw-root/build-root/winnt/lib/libmenuw.a
/usr/bin/install -c -m 644 ../lib/libmenuw.a /mingw-root/build-root/winnt/lib/libmenuw.a
installing ./eti.h in /mingw-root/build-root/winnt/include/ncursesw
installing ./menu.h in /mingw-root/build-root/winnt/include/ncursesw
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/menu“ wird verlassen
( cd form && make DESTDIR="" RPATH_LIST="/mingw-root/build-root/winnt/lib" install )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/form“ wird betreten
installing ../lib/libformw.a as /mingw-root/build-root/winnt/lib/libformw.a
/usr/bin/install -c -m 644 ../lib/libformw.a /mingw-root/build-root/winnt/lib/libformw.a
installing ./form.h in /mingw-root/build-root/winnt/include/ncursesw
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/form“ wird verlassen
( cd test && make DESTDIR="" RPATH_LIST="/mingw-root/build-root/winnt/lib" install )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/test“ wird betreten
make[1]: Für das Ziel „install“ ist nichts zu tun.
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/test“ wird verlassen
( cd misc && make DESTDIR="" RPATH_LIST="/mingw-root/build-root/winnt/lib" install )
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/misc“ wird betreten
mkdir -p /mingw-root/build-root/winnt/share
mkdir -p /mingw-root/build-root/winnt/share/tabset
DESTDIR= \
prefix=/mingw-root/build-root/winnt \
exec_prefix=/mingw-root/build-root/winnt \
bindir=/mingw-root/build-root/winnt/bin \
top_srcdir=.. \
srcdir=. \
datadir=/mingw-root/build-root/winnt/share \
ticdir=/mingw-root/build-root/winnt/share/terminfo \
source=terminfo.tmp \
cross_compiling=yes \
/bin/sh ./run_tic.sh
** Building terminfo database, please wait...
Running /usr/bin/tic to install /mingw-root/build-root/winnt/share/terminfo ...

        You may see messages regarding extended capabilities, e.g., AX.
        These are extended terminal capabilities which are compiled
        using
                tic -x
        If you have ncurses 4.2 applications, you should read the INSTALL
        document, and install the terminfo without the -x option.

ncurses 6.3.20211021
1788 entries written to /mingw-root/build-root/winnt/share/terminfo
** built new /mingw-root/build-root/winnt/share/terminfo
** sym-linked /mingw-root/build-root/winnt/lib/terminfo for compatibility
installing std
installing stdcrt
installing vt100
installing vt300
finished install.data
/usr/bin/install -c ncurses-config /mingw-root/build-root/winnt/bin/ncursesw6-config
installing ncursesw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/ncursesw.pc': Permission denied
installing ncursesw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/ncursesw.pc': Permission denied
installing panelw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/panelw.pc': Permission denied
installing menuw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/menuw.pc': Permission denied
installing formw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/formw.pc': Permission denied
installing ncurses++w.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/ncurses++w.pc': Permission denied
make[1]: *** [Makefile:150: install] Fehler 1
make[1]: Verzeichnis „/mingw-root/build-root/ncurses-6.3/misc“ wird verlassen
make: *** [Makefile:138: install] Fehler 2
```

Reply:
I've run into this file testing.  It's trying to install the pkg-config files for the mingw64 version to the global system because that's where pkg-config tells it to.  pkg-config should be disabled. I'll take a look at it, could attach your config.log?

Reply:
[config.log for ncurses](https://github.com/htop-dev/htop/files/9930220/config.log)

In my case it's properly installing the files in the proper platform directory (see start of the `make install` output), but further down it operating on files of the global system. With a prefix set this should not happen.

Reply:
These are the errors I'm seeing:

```
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/ncursesw.pc': Permission denied
installing ncursesw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/ncursesw.pc': Permission denied
installing panelw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/panelw.pc': Permission denied
installing menuw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/menuw.pc': Permission denied
installing formw.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/formw.pc': Permission denied
installing ncurses++w.pc
/usr/bin/install: Failed to remove '/usr/lib/x86_64-linux-gnu/pkgconfig/ncurses++w.pc': Permission denied
make[1]: *** [Makefile:150: install] Fehler 1
```

This is trying to install the pkg-config files for mingw64, but it's getting the location to put those files from running `pkg-config` instead basing it on the prefix. Getting pkg-config setup properly just seems unnecessary. Looking at your `config.log`  the wrong options are being given to `configure` . It should be disabling pkg-config stuff from being built (and then installed).  e.g., `--enable-pc-files` should be the opposite. 

Cleaning up the ncurses options along with testing that it was building binaries that would run properly on Windows was the last thing I did before my git adventure. I think I might have dropped a commit with the final version of the flags. I'll look at it and fix it.

Reply:
Was able to take a quick look. The `build_ncurses.sh` script and the configure options in the instructions are out of sync, with the script having the correct ones (from a quick look).

The scripts have been handy when I've been working, but would it be best to remove them?  If only to remove the possibility of them being out of sync with the documentation?

Reply:
I only looked at the instructions earlier when I tested this.

Given that documentation should usually not contain executables I'd suggest to remove those scripts.